spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py

```diff
@@ -52,37 +61,82 @@ def get_parser():
-            ax.scatter(x[i], y[i])
+            ax.scatter(x[i], y[i], color=colors[bvals[i]], alpha=0.7)
+    # plt.axis('equal')
     plt.xlabel(xlabel)
     plt.ylabel(ylabel)
-    plt.xlim([-1, 1])
-    plt.ylim([-1, 1])
+    plt.xlim([-max(bvals), max(bvals)])
+    plt.ylim([-max(bvals), max(bvals)])
     plt.grid()
 
+
+def create_custom_legend(fig, shell_colors, bvals):
+    """
+    Create single custom legend for whole figure
+    :param fig: figure the legend will be created for
+    :param shell_colors: dict with b-values and colors
+    :param bvals: ndarray with all b-values
+    :return:
+    """
+
+    # count number of bvals for individual shells
+    bvals_per_shell = pd.Series(bvals[bvals > bzero]).value_counts()
+
+    # Create single custom legend for whole figure with several subplots
+    # Loop across legend elements
+    lines = [Line2D([0], [0], color=color, marker='o', markersize=8, alpha=0.7, linestyle='')
+             for color in shell_colors.values()]
+    labels = [f'b-values = {shell} (n = {bvals_per_shell[shell]})'
+              for shell in shell_colors.keys()]
+
+    # Insert legend below subplots, NB - this line has to be run after the plt.tight_layout() which is called before
+    # function call
+    legend = fig.legend(lines, labels, loc='lower left', bbox_to_anchor=(0.1, 0),
+                        bbox_transform=plt.gcf().transFigure, ncol=len(lines), fontsize=10)
+    # Change box's frame color to black
+    frame = legend.get_frame()
+    frame.set_edgecolor('black')
+    # tight layout of whole figure and shift master title up
+    fig.tight_layout()
+    fig.subplots_adjust(top=0.88, bottom=0.1)
+
+
 # MAIN
 # ==========================================================================================
 
-
 def main(argv=None):
     parser = get_parser()
     arguments = parser.parse_args(argv)
     verbose = arguments.v
     set_global_loglevel(verbose=verbose)
 
     fname_bvecs = arguments.bvec
-
-    # Read bvecs
-    bvecs = read_bvals_bvecs(fname_bvecs, None)
-    bvecs = bvecs[0]
+    fname_bvals = arguments.bval
+    # Read bvals and bvecs files (if arguments.bval is not passed, bvals will be None)
+    bvals, bvecs = read_bvals_bvecs(fname_bvals, fname_bvecs)
     # if first dimension is not equal to 3 (x,y,z), transpose bvecs file
-    if not bvecs.shape[0] == 3:
+    if bvecs.shape[0] != 3:
         bvecs = bvecs.transpose()
-    x, y, z = bvecs[0], bvecs[1], bvecs[2]
+    # if bvals file was not passed, create dummy unit bvals array (necessary fot scatter plots)
+    if bvals is None:
+        bvals = np.repeat(1, bvecs.shape[1])
+    # multiply unit b-vectors by b-values
+    x, y, z = bvecs[0] * bvals, bvecs[1] * bvals, bvecs[2] * bvals
+
+    # Set different color for each shell (bval)
+    shell_colors = {}
+    # Create iterator with different colors from brg palette
+    colors = iter(cm.brg(np.linspace(0, 1, len(np.unique(bvals)))))
```

Comment:
Nice!

I tried this out with >6 unique bvals (just manually editing the bvals.txt file in sct_example_data), and I noticed that because the colors in the BRG colormap are a 3-color gradient, the intermediate colors were a little hard for me to tell apart:

![image](https://user-images.githubusercontent.com/16181459/118739332-36ea1080-b817-11eb-8445-872194332dc1.png)

I looked at [the matplotlib colormap list](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390/files#) and tried out a few other colomaps. `cm.nipy_spectral` felt the most distinct to me, but this is a bit personal preference too.

![image](https://user-images.githubusercontent.com/16181459/118739460-87616e00-b817-11eb-8d4f-aae448a81740.png)

Reply:
@joshuacwnewton Thanks for this test with >6 unique bvals! Indeed, the workaround with `linspace` has this disadvantage that color palette is changing unpredictably based on number of unique bvals. 

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d68efd6b68951cb54d3303b9767c7ac85458455d/spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py#L134

I will try to figure out some more elegant solution. But usually, common spherical multi-shell DWI sequences have up to 3 unique bvalues. (but of course, some [DSI q-space](https://www.researchgate.net/profile/Jian-Cheng/publication/230569009/figure/fig3/AS:437671634968579@1481360457930/Several-kinds-of-sampling-in-q-space-The-black-dot-in-q-0-0-0T-is-the-baseline.png) can have much more bvalues. So it is question how much time to invest to this thing 😄)

Reply:
Actually, now that I think about it, I should probably ask: Is it even important to tell the b-values apart? 

Maybe I misunderstood the purpose of the image. Since the bvals are ordered in the legend by their magnitude, it's true that the first colormap does effectively convey the increase (from low bvals to high bvals).

Reply:
What do you think of something like this, @valosekj? https://stackoverflow.com/questions/61084381/create-gradient-legend-matplotlib

Reply:
@joshuacwnewton The purpose of the image (at least for me) is to display how individual diffusion directions (bvecs file) and individual diffusion weighting (bvals file) are distributed over the q-space. Because from these two text files (bvals and bvecs) I don't get this information easily.
Important (again for me) is to be able to easily distinguish between individual shells (b-values).

Reply:
> Important (again at least for me) is to be able to easily distinguish between individual shells (b-values).

I appreciate the clarification! In that case, do you think switching the colormap to `cm.nipy_spectral` might help to make the shells more distinguishable? (see: figure in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390#discussion_r634835644)

Reply:
👍🏻 Color map switched in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390/commits/eef85738bde5a304736d8bacab68ac0dd92e945b